tools/qlog/src/lib.rs

```diff
@@ -258,6 +298,264 @@ impl Default for Qlog {
+            description,
+            summary,
+            traces: vec![trace],
+        };
+
+        QlogStreamer {
+            start_time,
+            writer,
+            qlog,
+            first_event: true,
+            first_frame: false,
+            writing_event: false,
+            writing_frames: false,
+            log_started: false,
+            log_finished: false,
+        }
+    }
+
+    /// Starts qlog streaming serialization.
+    ///
+    /// This writes out the JSON-serialized form of all information up to qlog
+    /// `Trace`'s array of `EventField`s. EventFields are separately appended
+    /// using functions that accept and `event::Event`.
+    pub fn write_start(&mut self) -> Result<()> {
+        // We can only start the log once.
+        if self.log_started {
+            return Err(Error::Done);
+        }
+
+        // A qlog contains a trace holding a vector of events that we want to
+        // serialize in a streaming manner. So at the start of serialization,
+        // take off all closing delimiters, and leave us in a state to accept
+        // new events.
+        // TODO: this unfortunately removes any events already existing in the
+        // trace, which might be undesirable.
+        match serde_json::to_string(&self.qlog) {
+            Ok(mut out) => {
+                if let Some(index) = out.find("\"events\":[") {
+                    out.truncate(index + "\"events\":[".len());
+                }
+
+                self.writer.as_mut().write_all(out.as_bytes())?;
+
+                self.log_started = true;
+            },
+
+            _ => return Err(Error::Done),
+        }
+
+        Ok(())
+    }
+
+    /// Finishes qlog streaming serialization.
+    ///
+    /// The JSON-serialized output has remaining close delimiters added.
+    /// After this is called, no more serialization will occur.
+    pub fn log_finish(&mut self) -> Result<()> {
+        if !self.log_started || self.log_finished {
+            return Err(Error::InvalidState);
+        }
+
+        self.writer.as_mut().write_all(b"]}]}")?;
+
+        self.log_finished = true;
+
+        self.writer.as_mut().flush()?;
+
+        Ok(())
+    }
+
+    /// Writes a JSON-serialized `EventField`s.
+    ///
+    /// Some qlog events can contain `QuicFrames`. If this is detected `true` is
+    /// returned and the streamer enters a frame-serialization mode that is only
+    /// concluded by `finish_frames()`. In this mode, attempts to log additional
+    /// events are ignored.
+    ///
+    /// If the event contains no array of `QuicFrames` return `false`.
+    pub fn write_event(&mut self, event: event::Event) -> Result<bool> {
+        if !self.log_started ||
+            self.log_finished ||
+            self.writing_event ||
+            self.writing_frames
+        {
+            return Err(Error::InvalidState);
+        }
+
+        let event_time = self.start_time.elapsed();
+
+        if let Some(t) = self.qlog.traces.get(0) {
```

Comment:
AFAICT this is always true, as the traces Vec will always have at least one element, so I think you can just do:
```rust
let t = self.qlog.traces[0];
```
and then avoid the additional block and indentation level.

Reply:
done